cpp/include/cudf/detail/aggregation/aggregation.hpp

```diff
@@ -27,6 +27,63 @@
 
 namespace cudf {
 namespace detail {
+
+class aggregation_finalizer;
+
+class compound_aggregation {
+ public:
+  virtual std::vector<aggregation> get_simple_aggregations(data_type col_type) const { return {}; }
+  virtual void finalize(aggregation_finalizer& finalizer) = 0;
+};
+
+// Forward declare compound aggregations.
+class mean_aggregation;
+class std_var_aggregation;
+class min_aggregation;
+class max_aggregation;
+
+// Visitor pattern
+class aggregation_finalizer {  // Declares the interface for the finalizer
+ public:
+  // Declare overloads for each kind of a agg to dispatch
+  virtual void Dispatch(min_aggregation& agg)     = 0;
+  virtual void Dispatch(max_aggregation& agg)     = 0;
+  virtual void Dispatch(mean_aggregation& agg)    = 0;
+  virtual void Dispatch(std_var_aggregation& agg) = 0;
+};
+
+/**
+ * @brief Derived class for specifying a min aggregation
+ */
+struct min_aggregation final : compound_aggregation, aggregation {
```

Comment:
Can `compound_aggregation` just inherit from `aggregation` instead of `min_aggregation` needing to inherit from both?

Reply:
Tried that. since `derived_aggregation` also inherits from `aggregation` too, both `aggregation` inheritance has to be virtual. That needs default constructor of `aggregation` to be defined. I assume, we don't want default constructor for `aggregation`.